### PR TITLE
feat: fetch new snapshots

### PR DESCRIPTION
**Pull request by pedrotambo**
This PR adds the functionality of reading snapshots of the new endpoint `/snapshots` that provides the active entities in multiple snapshots instead of a whole big one.
A `synchronizer` was added that is in charge of syncing the specified servers by deploying all the entities from snapshots and pointer-changes. It has mechanisms for retrying when things fail.
Also, with the possibility that a snapshot converges, if there are two servers (or more) with the same snapshot, it is downloaded only from one of them, and the entities inside are fetched from all the servers that have that snapshot as a load balancing mechanism.
The `synchronizer` maintains the servers in 3 states:
1. Bootstrap from snapshots: deploying the entities in the snapshots of that server.
2. Bootstrap from pointer-changes: deploying entities from pointer-changes between the last timestamp of its snapshots and now.
3. Syncing: deploying entities polling from pointer-changes.
When a server finishes the tasks in one state, goes to the next one. Retry mechanisms are done if something fails.

When all the initial servers to sync try to perform at least once with states (1) and (2), the initial bootstrap is considered as done, and the Catalysts startup. Note: if some server fails, it'll retry later but the initial bootstrap will finishes anyway.

### CODE REVIEW

File: src/client.ts [hunk 1]
```diff
@@ -3,11 +3,29 @@ import { metricsDefinitions } from './metrics'
...
 
-export async function getGlobalSnapshot(components: SnapshotsFetcherComponents, server: string, retries: number) {
-  const url = new URL(`${server}/snapshot`).toString()
+export async function getGlobalSnapshots(components: SnapshotsFetcherComponents, server: string, retries: number):
+  Promise<{
+    hash: string,
+    lastIncludedDeploymentTimestamp: number
+  }[]> {
+  try {
+    const newUrl = new URL(`${server}/snapshots`).toString()
```
Comment (agusaldasoro) [inline src/client.ts:12]: can we name this variable something like incrementalSnapshots or snapshotsRanges ?

File: src/client.ts [hunk 2]
```diff
@@ -3,11 +3,29 @@ import { metricsDefinitions } from './metrics'
...
+      lastIncludedDeploymentTimestamp: newSnapshot.timeRange.endTimestampSecs
+    }))
+  } catch {
+    components.logs.getLogger('snapshots-fetcher')
+      .info(`Error fetching new snapshots from ${server}. Will continue to fetch old one.`)
+  }
 
-  // TODO: validate response
-  return await components.downloadQueue.scheduleJobWithRetries(() => fetchJson(url, components.fetcher), retries)
+  const oldUrl = new URL(`${server}/snapshot`).toString()
```
Comment (agusaldasoro) [inline src/client.ts:27]: and this something like globalSnapshot or totalSnapshot?

File: src/client.ts [hunk 3]
```diff
@@ -3,11 +3,29 @@ import { metricsDefinitions } from './metrics'
...
+    lastIncludedDeploymentTimestamp: number
+  }[]> {
+  try {
+    const newUrl = new URL(`${server}/snapshots`).toString()
+    // TODO: validate response
+    const newSnapshots: {
+      hash: string,
+      timeRange: { initTimestampSecs: number, endTimestampSecs: number }
+    }[] = await components.downloadQueue.scheduleJobWithRetries(() => fetchJson(newUrl, components.fetcher), 1)
+    return newSnapshots.map(newSnapshot => ({
```
Comment (agusaldasoro) [inline src/client.ts:18]: what's the order of newSnapshots array? I'd like to process first the most recent snapshot batch, right?

File: test/getDeployedEntitiesStream.spec.ts [hunk 4]
```diff
@@ -18,11 +19,13 @@ test('getDeployedEntitiesStream', ({ components, stubComponents }) => {
   ]
   it('prepares the endpoints', () => {
     // serve the snapshots
-    components.router.get('/snapshot', async () => ({
-      body: {
+    components.router.get('/snapshots', async () => ({
```
Comment (agusaldasoro) [inline test/getDeployedEntitiesStream.spec.ts:22]: maybe we should leave a test for the old flow?

File: test/getDeployedEntitiesStream.spec.ts [hunk 5]
```diff
@@ -125,13 +122,199 @@ test('getDeployedEntitiesStream', ({ components, stubComponents }) => {
       { entityType: 'profile', entityId: 'Qm000013', localTimestamp: 13, authChain, pointers: ['0x1'] },
     ])
   })
+})
+
+test('fetches a stream without deleting the downloaded file', ({ components, stubComponents }) => {
```
Comment (agusaldasoro) [inline test/getDeployedEntitiesStream.spec.ts:127]: why do you need to not delete the file?

File: test/getDeployedEntitiesStream.spec.ts (hunk 4 above)
Comment (pedrotambo) [reply test/getDeployedEntitiesStream.spec.ts:22]: Yes, there are

File: test/getDeployedEntitiesStream.spec.ts (hunk 5 above)
Comment (pedrotambo) [reply test/getDeployedEntitiesStream.spec.ts:127]: Don't know. Was an old test, i just moved it

File: src/client.ts (hunk 1 above)
Comment (pedrotambo) [reply src/client.ts:13]: yes

File: src/client.ts (hunk 3 above)
Comment (pedrotambo) [reply src/client.ts]: Added

File: src/index.ts [hunk 6]
```diff
@@ -142,54 +142,64 @@ export async function* getDeployedEntitiesStream(
...
+          }
         }
-      }
-    } finally {
-      if (options.deleteSnapshotAfterUsage !== false) {
-        try {
-          await components.storage.delete([hash])
-        } catch (err: any) {
-          logs.error(err)
+        await components.processedSnapshotStorage.markSnapshotProcessed(hash, replacedSnapshotHashes)
```
Comment (menduz) [inline src/index.ts:187]: This line does not ensures that all the entities were "deployed" right? only sent to the next stage of the pipeline (p-queue), which will later deploy them.

Does this logic risk losing deployments? Assume that the queue fails or we restart the service while it is draining. Will those deployments be lost?

1. Read snapshot bafky1
2. Send all 1000 entities to the queue (yield)
3. Mark bafky1 as read
4. Deploy up to 101 entities from the queue
5. *SERVICE RESTART*
6. ~Read snapshot bafky1~ -> IGNORE, bafky1 was already processed
7. *899 deployments lost in the limbo*

File: src/index.ts [hunk 7]
```diff
@@ -142,54 +142,64 @@ export async function* getDeployedEntitiesStream(
...
+            logs.error(err)
+          }
         }
       }
     }
+    greatestProcessedTimestamp = lastIncludedDeploymentTimestamp > greatestProcessedTimestamp
+      ? lastIncludedDeploymentTimestamp : greatestProcessedTimestamp
   }
 
+  logs.info(`End processing snapshots. Greatest processed timestamp: ${greatestProcessedTimestamp}`)
```
Comment (menduz) [inline src/index.ts:202]: Prefer logs without string interpolations
```suggestion
  logs.info(`End processing snapshots.`, {greatestProcessedTimestamp})
```

**Review by menduz — COMMENTED** [review]
Looking good, few comments, none blocker but should be addressed at some time. Would you also add a migration guide for all the projects using the lib?

File: src/index.ts [hunk 8]
```diff
@@ -142,54 +142,64 @@ export async function* getDeployedEntitiesStream(
...
-        hash,
-        options.tmpDownloadFolder,
-        [options.contentServer],
-        new Map(),
-        options.requestMaxRetries,
-        options.requestRetryWaitTime
-      )
+  for (const snapshot of snapshots) {
+    const { hash, lastIncludedDeploymentTimestamp, replacedSnapshotHashes } = snapshot
+    logs.debug(`Processing snapshot: ${snapshot.hash} from ${options.contentServer}`)
```
[the remark this answers is not in the document]
Comment (pedrotambo) [reply src/index.ts:154]: Good to know 👍

File: src/index.ts (hunk 6 above)
Comment (pedrotambo) [reply src/index.ts:187]: This is correct. This logic risks losing deployments. 😭

File: src/index.ts [hunk 9]
```diff
@@ -142,54 +142,64 @@ export async function* getDeployedEntitiesStream(
...
+    // 2. for each snapshot, download the snapshot file if it contains deployments
+    //    in the range we are interested (>= genesisTimestamp)
+    if (
+      hash && lastIncludedDeploymentTimestamp && lastIncludedDeploymentTimestamp > genesisTimestamp &&
+      !(await components.processedSnapshotStorage.wasSnapshotProcessed(hash, replacedSnapshotHashes))) {
```
[the remark this answers is not in the document]
Comment (menduz) [reply src/index.ts:159]: it will happen at most a couple hundred times per day, in this case I'd optimize readable code

File: src/index.ts [hunk 10]
```diff
@@ -78,29 +118,15 @@ export async function downloadEntityAndContentFiles(
      * A proper fix needs to be added that validates and forces new deployments to include the files in content (even
      *  if no need to upload them again).
```
Comment (agusaldasoro) [inline src/index.ts:119]: Do we have this tracked somewhere?

File: src/index.ts [hunk 11]
```diff
@@ -24,6 +25,45 @@ if (parseInt(process.version.split('.')[0]) < 16) {
...
+  const allAvatars: any[] = entityMetadata.metadata?.avatars ?? []
+  const snapshots = allAvatars.flatMap(avatar => Object.values(avatar.avatar.snapshots ?? {}) as string[])
+    .filter(snapshot => !!snapshot)
+    .map(snapshot => {
+      const matches = snapshot.match(/^http.*\/content\/contents\/(.*)/)
+      return matches ? matches[1] : snapshot
+    })
+    .filter(snapshot => !entityMetadata.content || entityMetadata.content.find(content => content.hash === snapshot) === undefined)
+  if (snapshots.length > 0) {
+    logger.info(`Downloading snapshots ${snapshots} for fixing entity ${JSON.stringify(entityMetadata)}`)
```
Comment (agusaldasoro) [inline src/index.ts:51]:
```suggestion
    logger.info(`Downloading snapshots ${snapshots} for fixing profile entity ${JSON.stringify(entityMetadata)}`)
```

Note that these snapshots are not the same as the snapshots used for sync

File: src/index.ts (hunk 6 above)
Comment (agusaldasoro) [reply src/index.ts:217]: Aren't we storing the failed deployments to avoid that?

**Review by agusaldasoro — APPROVED** [review]

File: src/index.ts (hunk 6 above)
Comment (menduz) [reply src/index.ts:217]: the failed deployments are stored upon failure, not when added to the queue

Comment (pedrotambo) [reply src/index.ts:217]: Yes, but the restart could happen before a deployment is done (failed or not)

File: src/client.ts [hunk 12]
```diff
@@ -3,11 +3,35 @@ import { metricsDefinitions } from './metrics'
...
+    replacedSnapshotHashes?: string[]
+  }[]> {
+  try {
+    const incrementalSnapshotsUrl = new URL(`${server}/snapshots`).toString()
+    // TODO: validate response
+    const newSnapshots: {
+      hash: string,
+      timeRange: { initTimestamp: number, endTimestamp: number },
+      replacedSnapshotHashes?: string[]
+    }[] = await components.downloadQueue.scheduleJobWithRetries(() => fetchJson(incrementalSnapshotsUrl, components.fetcher), 1)
```
[the remark this answers is not in the document]
Comment (pedrotambo) [reply src/client.ts:19]: changed to 3 retries

File: src/index.ts (hunk 10 above)
Comment (pedrotambo) [reply src/index.ts:120]: i think so

File: src/client.ts [hunk 13]
```diff
@@ -3,11 +3,35 @@ import { metricsDefinitions } from './metrics'
...
-  const url = new URL(`${server}/snapshot`).toString()
+export async function getSnapshots(components: SnapshotsFetcherComponents, server: string, retries: number):
+  Promise<{
+    hash: string,
+    lastIncludedDeploymentTimestamp: number
+    replacedSnapshotHashes?: string[]
+  }[]> {
+  try {
+    const incrementalSnapshotsUrl = new URL(`${server}/snapshots`).toString()
+    // TODO: validate response
```
[the remark this answers is not in the document]
Comment (pedrotambo) [reply src/client.ts:14]: removed

File: src/index.ts (hunk 6 above)
Comment (pedrotambo) [reply src/index.ts]: This issue was resolved. When the snapshots-fetcher streams entities, it sends a callback that the Catalyst uses when the entity was effectively deployed or persisted as failed deployed. The batch deployer gets an element from the queue, effectively deploys or persists as failed deployment and at the end calls the callback sent by the snapshot-fetcher saying "effectively deployed an entity in this snapshot".

**Pull request closed without merging**